### PR TITLE
Add esm build

### PR DESCRIPTION
**Pull request by dagda1**
## Motivation

This PR adds an `esm` build to the existing commonjs build.

## Approach

- commonjs builds to dist/cjs
- esm builds to dist/esm
- there is a new tsconfig.dist-esm.json file that extends tsconfg.dist.json
```json
{
   "extends": "./tsconfig.cjs.json",
   "compilerOptions": {
    "outDir": "./dist/esm",
    "module": "ESNext"
  }
}
```

### CODE REVIEW

File: packages/core/package.json [hunk 1]
```diff
@@ -16,9 +17,10 @@
     "test": "mocha -r ts-node/register 'test/**/*.test.ts'",
     "mocha": "mocha -r ts-node/register",
     "build": "yarn prepack",
-    "prepack": "tsc --build tsconfig.dist.json"
+    "prepack": "copy ./package.json ./dist && tsc --build tsconfig.esm.json && tsc --build tsconfig.cjs.json"
   },
   "files": [
+    "package.json",
```
Comment (cowboyd) [inline packages/core/package.json]: I'm pretty sure this is always mandatory, and will always be included with the tarball

File: patches/typedoc+0.21.0.patch [hunk 2]
```diff
@@ -1,48 +0,0 @@
...
-     let packageTypes = null;
--    if (hasOwnProperty(packageJson, "main") &&
-+    if (hasOwnProperty(packageJson, "typeDocEntry") && 
-+        typeof packageJson.typeDocEntry == "string") {
-+        packageMain = packageJson.typeDocEntry;
-+    }
-+    else if (hasOwnProperty(packageJson, "main") &&
-         typeof packageJson.main == "string") {
-         packageMain = packageJson.main;
-     }
```
Comment (cowboyd) [inline patches/typedoc+0.21.0.patch]: Is this related? I thought that we needed this patch for the API docs /cc @minkimcello

**Review by cowboyd — APPROVED** [review]

File: packages/core/package.json [hunk 3]
```diff
@@ -13,11 +13,12 @@
     "directory": "packages/core"
   },
   "scripts": {
+    "clean": "rimraf *.tsbuildinfo dist",
     "lint": "eslint '{src,bin,test}/**/*.ts'",
     "test": "mocha -r ts-node/register 'test/**/*.test.ts'",
     "mocha": "mocha -r ts-node/register",
     "build": "yarn prepack",
-    "prepack": "copy \"package.json\" \"dist\"  && tsc --build tsconfig.esm.json && tsc --build tsconfig.cjs.json"
+    "prepack": "yarn clean && tsc --build --clean && copy \"package.json\" \"dist\"  && tsc --build tsconfig.esm.json && tsc --build tsconfig.cjs.json"
```
Comment (cowboyd) [inline packages/core/package.json]: Why is this necessary? Won't it circumvent the optimizations we have by using references?

Comment (dagda1) [reply packages/core/package.json]: It was just something I was trying in desperation to get this passing on windows.  It is removed now
